### PR TITLE
Add support for visionfive2 platform

### LINKED ISSUE #104: Platform for VisionFive2 board
Add support for the physical board [VisionFive2](https://www.starfivetech.com/en/site/boards) from starfive. 
We need to create a new Platform implementation to abstract technical properties of the board and have a proper interface with the hardware.

- [x] List the useful addresses (CLINT, serial, ...)
- [x] Can we power off board from software ?
- [x] Add a way to choose the platform from the config file

### PR DESCRIPTION
**Pull request by NoeTerrier**
Add platform implementation for VisionFive2 board and a way to select the platform in the Cargo.toml file.

Closes #104

### CODE REVIEW

File: src/platform/mod.rs [hunk 1]
```diff
@@ -11,9 +12,14 @@ use crate::{config, device, logger};
 
 /// Export the current platform.
 /// For now, only QEMU's Virt board is supported
+#[cfg(not(feature = "platform_visionfive2"))]
 pub type Plat = virt::VirtPlatform;
 
+#[cfg(feature = "platform_visionfive2")]
+pub type Plat = visionfive2::VisionFive2Platform;
```
Comment (CharlyCst) [inline src/platform/mod.rs:19]: The `platform_virt` feature is never used, we can remove it and assume it is the default.

File: src/platform/mod.rs [hunk 2]
```diff
@@ -34,12 +40,15 @@ pub trait Platform {
     fn get_max_valid_address() -> usize;
 
     const HAS_S_MODE: bool = config::VCPU_S_MODE;
+    const NB_HARTS: usize = usize::MAX;
```
Comment (CharlyCst) [inline src/platform/mod.rs:43]: Do you have to provide a default here? Would that work or does the compiler complain?

```suggestion
    const NB_HARTS: usize;
```

File: src/platform/mod.rs [hunk 3]
```diff
@@ -34,12 +40,15 @@ pub trait Platform {
...
 }
 
 pub fn init() {
     Plat::init();
     logger::init();
 
+    log::info!("{:?}", Plat::name());
+
```
Comment (CharlyCst) [inline src/platform/mod.rs:51]: We can maybe move this log into `main.rs`, we already have a bunch of logs at boot time and this one could go there.

Maybe add some context, like:

```rs
log::info!("Platform: {:?}", Plat::name());
```

File: src/platform/virt.rs [hunk 4]
```diff
@@ -36,6 +37,10 @@ static VIRT_CLINT: VirtClint = VirtClint::new(&CLINT_MUTEX);
 pub struct VirtPlatform {}
 
 impl Platform for VirtPlatform {
+    fn name() -> &'static str {
+        "Virtual"
```
Comment (CharlyCst) [inline src/platform/virt.rs:41]: Maybe "QEMU virt", it's how I have seen it called elsewhere (and `virt` is the official machine name in QEMU, we invoke it with `--machine virt`).

File: src/platform/visionfive2.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+use super::Platform;
+use crate::arch::{Arch, Architecture};
+use crate::device::{self, VirtClint};
+use crate::driver::ClintDriver;
+
+// —————————————————————————— Platform Parameters ——————————————————————————— //
+
+const SERIAL_PORT_BASE_ADDRESS: usize = 0x10000000;
+const MIRALIS_START_ADDR: usize = 0x80000000;
```
Comment (CharlyCst) [inline src/platform/visionfive2.rs:17]: This one will be different, maybe @Wesdcv can give us the boot address :)

File: src/platform/visionfive2.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    fn exit_failure() -> ! {
+        loop {
+            Arch::wfi();
+            hint::spin_loop();
+        }
+    }
+
+    fn load_firmware() -> usize {
+        // We directly load the firmware from QEMU, nothing to do here.
```
Comment (CharlyCst) [inline src/platform/visionfive2.rs:78]: Let's remove the comment, it doesn't make sense here.

```suggestion
```

**Review by CharlyCst — COMMENTED** [review]
Thanks for the changes! I just have a few minor comments and we are all good :)

**Comment by CharlyCst** [comment]
Regarding the CI it fails because there is a merge commit.

In general try to use `git rebase main` rather than merging main into your branch. If you merge you create cycles which makes it hard to reason about the history. Here is the history with this branch for illustration:

```
@  zlmypmpn charly.castes@epfl.ch 2024-07-19 00:20:42 c369fd86
│  (empty) (no description set)
│ ◉  mrpsmtup noeterrier@proton.me 2024-07-18 16:17:56 104-visionfive2-platform@origin 24f7ce0e
╭─┤  (empty) Merge branch 'main' into 104-visionfive2-platform
◉ │  rqrupxnm charly.castes@epfl.ch 2024-07-18 09:48:45 105-support-for-mip-csr@origin main HEAD@git 2482e808
│ │  Display QEMU version in the CI
◌ │  (elided revisions)
│ ◉  yukvpzsu noeterrier@proton.me 2024-07-18 16:15:07 65105284
├─╯  Add support for visionfive2 platform
◉  mqkvqoyz charly.castes@epfl.ch 2024-07-17 23:22:12 e00e173e
│  Rename Mirage to Miralis
```

Feel free to ask if in doubt, and don't worry about that one I will linearise the history before merging ^^

File: src/platform/mod.rs (hunk 2 above)
Comment (NoeTerrier) [reply src/platform/mod.rs:43]: I can remove it, but I need to define it at compile time in VirtPlatform (usize::MAX ?)

**Review by CharlyCst — COMMENTED** [review]
Looks good to me, thanks for incorporating the changes :)

There seems to be only minor issues left (format check is failing), I'll handle those and merge the PR.

**Pull request merged**